kbs/src/token/jwk.rs

```diff
@@ -0,0 +1,166 @@
+    match Url::parse(p) {
+        Err(e) => Err(JwksGetError::SourcePath(e.to_string())),
+        Ok(mut url) if url.scheme() == "https" => {
+            url.set_path(OPENID_CONFIG_URL_SUFFIX);
+
+            let oidc = get(url.as_str())
+                .await
+                .map_err(|e| JwksGetError::SourceAccess(e.to_string()))?
+                .json::<OpenIDConfig>()
+                .await
+                .map_err(|e| JwksGetError::SourceDeserializeJson(e.to_string()))?;
+
+            let jwkset = get(oidc.jwks_uri)
+                .await
+                .map_err(|e| JwksGetError::SourceAccess(e.to_string()))?
+                .json::<jwk::JwkSet>()
+                .await
+                .map_err(|e| JwksGetError::SourceDeserializeJson(e.to_string()))?;
+
+            Ok(jwkset)
+        }
+        Ok(url) if url.scheme() == "file" && Path::new(url.path()).exists() => {
+            let file = File::open(url.path())
+                .map_err(|e| JwksGetError::SourceAccess(format!("open {}: {}", url.path(), e)))?;
+            let reader = BufReader::new(file);
+
+            serde_json::from_reader(reader)
+                .map_err(|e| JwksGetError::SourceDeserializeJson(e.to_string()))
+        }
+        Ok(url) => Err(JwksGetError::SourcePath(format!(
+            "unsupported scheme {} (must be either file or https) or missing file path",
+            url.scheme()
+        ))),
+    }
+}
+
+impl JwkAttestationTokenVerifier {
+    pub async fn new(config: &AttestationTokenVerifierConfig) -> Result<Self> {
+        let trusted_certs = match &config.trusted_certs_paths {
+            Some(paths) => {
+                let mut keyset = jwk::JwkSet { keys: Vec::new() };
+
+                for path in paths.iter() {
+                    match get_jwks_from_file_or_url(path).await {
+                        Ok(mut jwkset) => keyset.keys.append(&mut jwkset.keys),
+                        Err(e) => log::warn!("error getting JWKS: {:?}", e),
+                    }
+                }
+
+                match keyset.keys.len() {
+                    0 => None,
+                    _ => Some(keyset),
+                }
+            }
+            None => None,
+        };
+
+        Ok(Self { trusted_certs })
+    }
+}
+
+#[async_trait]
+impl AttestationTokenVerifier for JwkAttestationTokenVerifier {
+    async fn verify(&self, token: String) -> Result<String> {
+        let Some(keyset) = &self.trusted_certs else {
+            bail!("Cannot verity token since trusted JWK Set is empty");
+        };
+
+        let kid = decode_header(&token)
+            .context("Failed to decode attestation token header")?
+            .kid
+            .ok_or(anyhow!("Failed to decode kid in the token header"))?;
+
+        let key = keyset
+            .find(&kid)
+            .ok_or(anyhow!("Failed to find kid in trusted certificates"))?;
+
+        let alg = Algorithm::from_str(key.common.key_algorithm.unwrap().to_string().as_str())?;
```

Comment:
Do we know that this key algorithm will always be there? You might split this line into two. It's a bit of a mouthful.

Reply:
yeah better check for `Option<KeyAlgorithm>` is justified

Reply:
updated.